In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('weather_data.csv')

In [3]:
df.head()

day  temperature  windspeed  event
0  1/1/2017           32          6   Rain
1  1/2/2017           35          7  Sunny
2  1/3/2017           28          2   Snow
3  1/4/2017           24          7   Snow
4  1/5/2017           32          4   Rain

## Pandas Series

A Pandas Series is like a column in a table.

It is a one-dimensional array holding data of any type.

In [4]:
a = [1, 7, 2]

myvar = pd.Series(a)

In [5]:
myvar

0    1
1    7
2    2
dtype: int64

### Labels

If nothing else is specified, the values are labeled with their index number. First value has index 0, second value has index 1 etc.

This label can be used to access a specified value.

In [6]:
print(myvar[0])

1


<b> Create Labels 

In [7]:
a = [1, 7, 2]

myvar = pd.Series(a, index = ["x", "y", "z"])

In [8]:
print(myvar)

x    1
y    7
z    2
dtype: int64


### Key / Value Objects as Series

You can also use a key/value object, like a dictionary, when creating a Series.

In [9]:
calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories)

print(myvar)

day1    420
day2    380
day3    390
dtype: int64


Create a Series using only data from "day1" and "day2"

In [10]:
calories = {"day1": 420, "day2": 380, "day3": 390}

myvar = pd.Series(calories, index = ["day1", "day2"])

In [11]:
myvar

day1    420
day2    380
dtype: int64

## DataFrames

Data sets in Pandas are usually multi dimensional tables, called DataFrames.

Series is like a column, a DataFrame is the whole table.

In [12]:
data = {
    "calories": [420, 380, 390],
    "duration": [50, 40, 45]
}

myvar = pd.DataFrame(data)

In [13]:
print(myvar)

   calories  duration
0       420        50
1       380        40
2       390        45


### Locate Row

As you can see from the result above, the DataFrame is like a table with rows and columns. 

Pandas use the `loc` attribute to return one or more specified rows(s)

In [14]:
## 

print(df.loc[0])

day            1/1/2017
temperature          32
windspeed             6
event              Rain
Name: 0, dtype: object


In [15]:
df

day  temperature  windspeed  event
0  1/1/2017           32          6   Rain
1  1/2/2017           35          7  Sunny
2  1/3/2017           28          2   Snow
3  1/4/2017           24          7   Snow
4  1/5/2017           32          4   Rain
5  1/6/2017           31          2  Sunny

In [19]:
## Retreive 2 rows from dataset

print(df.loc[[0,1]])

        day  temperature  windspeed  event
0  1/1/2017           32          6   Rain
1  1/2/2017           35          7  Sunny


#### Named Indexes

With the `index` argument, you can name your own indexes.

In [20]:
data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

df = pd.DataFrame(data, index = ["day1", "day2", "day3"])

In [21]:
df

calories  duration
day1       420        50
day2       380        40
day3       390        45

In [22]:
## Located names indexes

print(df.loc["day2"])

calories    380
duration     40
Name: day2, dtype: int64


In [23]:
df = pd.read_csv('data.csv')

In [24]:
df.to_string()

'                Make                                Model  Year                              Engine Fuel Type  Engine HP  Engine Cylinders Transmission Type      Driven_Wheels  Number of Doors                                         Market Category Vehicle Size        Vehicle Style  highway MPG  city mpg  Popularity     MSRP\n0                BMW                           1 Series M  2011                   premium unleaded (required)      335.0               6.0            MANUAL   rear wheel drive              2.0                   Factory Tuner,Luxury,High-Performance      Compact                Coupe           26        19        3916    46135\n1                BMW                             1 Series  2011                   premium unleaded (required)      300.0               6.0            MANUAL   rear wheel drive              2.0                                      Luxury,Performance      Compact          Convertible           28        19        3916    40650\n2              

.

.

.

.

# Data Cleaning

Data cleaning means fixing bad data in your data set.

Bad data could be:

- Empty cells
- Data in wrong format
- Wrong data
- Duplicates

In [51]:
data = pd.read_csv('dirtydata.csv')

In [52]:
data

Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'    100       120     300.0
18        45  '2020/12/18'     90       112       NaN
19        60  '2020/12/19'    103       123     323.0
20        45  '2020/12/20'     97       125     243.0
21        60  '2020/12/21'    108       131     364.2
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0
27        60  '2020/12/27'     92       118     241.0
28        60  '2020/12/28'    103       132       NaN
29        60  '2020/12/29'    100       132     280.0
30        60  '2020/12/30'    102       129     380.3
31        60  '2020/12/31'     92       115     243.0

### Empty Cells

In [53]:
new_df = data.dropna()

In [54]:
print(new_df.to_string)

<bound method DataFrame.to_string of     Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120  

# Data of Wrong Format

Cells with data of wrong format can make it difficult, or even impossible, to analyze data.

To fix it, you have two options: remove the rows, or convert all cells in the columns into the same format.

### Convert into a correct Format

In our Data Frame, we have two cells with the wrong format. Check out row 22 and 26, the 'Date' column should be a string that represents a date:

In [56]:
data.loc[22: 26]

Duration          Date  Pulse  Maxpulse  Calories
22        45           NaN    100       119     282.0
23        60  '2020/12/23'    130       101     300.0
24        45  '2020/12/24'    105       132     246.0
25        60  '2020/12/25'    102       126     334.5
26        60      20201226    100       120     250.0

Let's try to convert all cells in the 'Date' column into dates.

Pandas has `to_datetime()` method for this:

In [57]:
data.columns

Index(['Duration', 'Date', 'Pulse', 'Maxpulse', 'Calories'], dtype='object')

In [58]:
data['Date'] = pd.to_datetime(data['Date'])

In [59]:
data

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
22        45        NaT    100       119     282.0
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

.

## Removing Rows

In [60]:
data.dropna(subset = ['Date'], inplace = True)

In [61]:
data

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

.

# Fixing Wrong Data

"Wrong data" does not have to be "empty cells" or "wrong format", it can just be wrong, like if someone registered "199" instead of "1.99".

Sometimes you can spot wrong data by looking at the dataset, because you have an expectation of what it should be.

If you take a look at our data set, you can see that in row 7, the duration is 450, but for all the other rows the duration is between 30 and 60.

It doesn't have to be wrong, but taking in consideration that this is the data set of someone's workout sessions, we conclude with the fact that this person did not work out in 450 minutes.

## Replacing Values

One way to fix wrong values is to replace them with something else.

In our example, it is most likely a typo, and the value should be "45" instead of "450", and we could just insert "45" in row 7:

In [64]:
data.loc[7,"Duration"] = 45

In [65]:
data

Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7         45 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90       112       NaN
19        60 2020-12-19    103       123     323.0
20        45 2020-12-20     97       125     243.0
21        60 2020-12-21    108       131     364.2
23        60 2020-12-23    130       101     300.0
24        45 2020-12-24    105       132     246.0
25        60 2020-12-25    102       126     334.5
26        60 2020-12-26    100       120     250.0
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

To replace wrong data in a larger data sets you can create some rules, e.g set some boundaries for legal values, and replace any values that are outside of the boundaries

In [66]:
for x in data.index:
    if df.loc[x, "Duration"] > 120:
        df.loc[x, "Duration"] = 120

### or Removing Rows

In [67]:
for x in df.index:
    if df.loc[x, "Duration"] > 120:
        df.drop(x, inplace = True)

.

# Pandas Removing Duplicates

Duplicate rows are rows that have been registered more than one time.



In [70]:
print(df.duplicated())

0        False
1        False
2        False
3        False
4        False
         ...  
11909    False
11910    False
11911    False
11912    False
11913    False
Length: 11914, dtype: bool


In [71]:
df.drop_duplicates(inplace = True)

In [72]:
df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
11909    False
11910    False
11911    False
11912    False
11913    False
Length: 11199, dtype: bool

.

# Data Correlations

In [73]:
data.corr()

Duration     Pulse  Maxpulse  Calories
Duration  1.000000 -0.090626 -0.304610  0.330900
Pulse    -0.090626  1.000000  0.269000  0.510361
Maxpulse -0.304610  0.269000  1.000000  0.352529
Calories  0.330900  0.510361  0.352529  1.000000

In [76]:
data.head()

Duration       Date  Pulse  Maxpulse  Calories
0        60 2020-12-01    110       130     409.1
1        60 2020-12-02    117       145     479.0
2        60 2020-12-03    103       135     340.0
3        45 2020-12-04    109       175     282.4
4        45 2020-12-05    117       148     406.0

In [78]:
data.head()

Duration       Date  Pulse  Maxpulse  Calories
0        60 2020-12-01    110       130     409.1
1        60 2020-12-02    117       145     479.0
2        60 2020-12-03    103       135     340.0
3        45 2020-12-04    109       175     282.4
4        45 2020-12-05    117       148     406.0

In [79]:
data.tail()

Duration       Date  Pulse  Maxpulse  Calories
27        60 2020-12-27     92       118     241.0
28        60 2020-12-28    103       132       NaN
29        60 2020-12-29    100       132     280.0
30        60 2020-12-30    102       129     380.3
31        60 2020-12-31     92       115     243.0

# Pandas apply()

Pandas.apply allows the users to pass a function and apply it on every single value of the Pandas series. It comes as a huge improvement for the pandas library as this function helps to segregate data according to the conditions required due to which it is efficiently used in data science and machine learning.

Example 1:

In [93]:
s = pd.read_csv("stock.csv")

## squeeze basically creates into Series

In [94]:
## squeeze basically creates into Series
s

Stock Price
0           50.12
1           54.10
2           54.65
3           52.38
4           52.95
...           ...
3007       772.88
3008       771.07
3009       773.18
3010       771.61
3011       782.22

[3012 rows x 1 columns]

In [95]:
s = pd.read_csv("stock.csv", squeeze = True)


In [96]:
s

0        50.12
1        54.10
2        54.65
3        52.38
4        52.95
         ...  
3007    772.88
3008    771.07
3009    773.18
3010    771.61
3011    782.22
Name: Stock Price, Length: 3012, dtype: float64

In [97]:
def fun(num):
    
    if num < 200:
        return "Low"
    
    elif num >= 200 and num < 400:
        return "Normal"
    else:
        return "High"

In [98]:
new = s.apply(fun)

In [100]:
print(new.head(3))

0    Low
1    Low
2    Low
Name: Stock Price, dtype: object


In [102]:
## printing elements somewhere near the middle of series

print(new[1400], new[1500], new[1600])



Normal Normal Normal


In [103]:
## printing last 3 elements
print(new.tail(3))

3009    High
3010    High
3011    High
Name: Stock Price, dtype: object


.

<b> Example 2:

In [104]:
s = pd.read_csv("stock.csv", squeeze = True)

In [105]:
## adding 5 to each value

new = s.apply(lambda num: num + 5)

In [106]:
print(s.head(), '\n', new.head())

0    50.12
1    54.10
2    54.65
3    52.38
4    52.95
Name: Stock Price, dtype: float64 
 0    55.12
1    59.10
2    59.65
3    57.38
4    57.95
Name: Stock Price, dtype: float64


.

# Pandas.map()

pandas.map() is used to map values from two series having one column same. For mapping two series, the last column of the first series should be same as index column of the second series, also the values should be unique

In [107]:
data.head()

Duration       Date  Pulse  Maxpulse  Calories
0        60 2020-12-01    110       130     409.1
1        60 2020-12-02    117       145     479.0
2        60 2020-12-03    103       135     340.0
3        45 2020-12-04    109       175     282.4
4        45 2020-12-05    117       148     406.0

In [109]:
calories = pd.read_csv('data.csv', usecols = ["Calories"],squeeze = True)

## usecol is used to use selected columns
## index_col is used to make passed column as index

types = pd.read_csv('data.csv', index_col = "Calories", squeeze = True)

ValueError: Usecols do not match columns, columns expected but not found: ['Calories']